bin/oneoff/back_populate_user_levels_level_source_id

```diff
@@ -50,28 +50,42 @@ batch_size = ARGV[1].to_i
 ActiveRecord::Base.record_timestamps = false
 
 begin
-  # Read the contents of the file, saving them to an array so that we can batch
-  # the updates in a transaction.
-  update_array = []
+  # Read the contents of the file, saving them to a hash so that we can update
+  # all UserLevel's associated with a fixed LevelSource at once.
+  puts "Building update_hash..."
+  update_hash = Hash.new {|hash, key| hash[key] = []}
   CSV.foreach(filename) do |line|
     user_levels_id = line[0].to_i
     level_source_id = line[1].to_i
-    update_array << [user_levels_id, level_source_id]
+    update_hash[level_source_id] << user_levels_id
   end
 
-  puts "Found #{update_array.length} updates to process..."
+  puts "Built update_hash with #{update_hash.size} distinct level_source_id's."
 
-  # Process the array in batches, so that the transactions are appropriate
-  # sized.
-  batch_index = 0
-  update_array.each_slice(batch_size) do |update_array_slice|
-    puts "Updating #{batch_index} to #{batch_index + batch_size} of #{update_array.length}..."
-    batch_index += batch_size
+  # Transform the hash into an array of hashes, each with batch_size many keys,
+  # for batching in transactions.
+  count = 0
+  update_hash_batches = []
+  current_hash = {}
+  update_hash.each_pair do |level_source_id, user_levels_id_array|
+    if count % batch_size == 0
+      update_hash_batches << current_hash
+      current_hash = {}
+    end
+    current_hash[level_source_id] = user_levels_id_array
+
+    count += 1
+  end
```

Comment:
This could be done more efficiently by converting the hash to an array and use `each_slice`, as aoby suggested offline. However, as this is what is running in production, I'm going to keep it as is (at least for now).